### PR TITLE
[IA-3750] Convert Analysis Modal to TypeScript.

### PR DESCRIPTION
**Pull request by n8landolt**
https://broadworkbench.atlassian.net/browse/IA-3891

# Description

We want to convert AnalysisModal to typescript to get the benefits of type safety.

This PR far is the first step for the ticket - get the type script version of the file functioning.

[IA-3891]: https://broadworkbench.atlassian.net/browse/IA-3891?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by n8landolt** [comment]
One challenge I encountered was with the loading spinners and effectively updating the state.
In our original code, AnalysisModal would trigger a loading spinner on the file screen behind it. However, when using useAnalysisFiles, updating that loading seems more difficult.

How do we want to handle these loading spinners/busy state? In this PR I added a busy/setBusy for AnalysisModal - the UX seems a bit clearer than the current implementation, but I'm not sure my approach is what we want long term.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 1]
```diff
@@ -18,45 +18,73 @@ import { FormLabel } from 'src/libs/forms'
...
 
-    const currentRuntime = getCurrentRuntime(runtimes)
+    const currentRuntime: any = getCurrentRuntime(runtimes)
     const currentDisk = getCurrentPersistentDisk(runtimes, persistentDisks)
     const currentRuntimeTool = getToolFromRuntime(currentRuntime)
-    const currentApp = toolLabel => getCurrentApp(getAppType(toolLabel))(apps)
+    const currentApp: any = toolLabel => getCurrentApp(getAppType(toolLabel))(apps)
+
+    //TODO: Bring in as props from Analyses
+    const { loadedState, create, pendingCreate }: AnalysisFileStore = useAnalysisFiles()
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: I am thinking that once the parent component is doing this useAnalysisFiles logic, this modal can be changed to just accept the analysis files array as a prop, removing the need to handle null or loadedState variants here....

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 2]
```diff
@@ -312,18 +360,19 @@ export const AnalysisModal = withDisplayName('AnalysisModal')(
...
                   [isRStudio, () => baseRmd])
                 const fullAnalysisName = `${analysisName}.${fileExt}`
-                !!googleProject ?
-                  await Ajax().Buckets.analysis(googleProject, bucketName, fullAnalysisName, toolLabel).create(contents) :
-                  await Ajax().AzureStorage.blob(workspaceId, fullAnalysisName).create(contents)
+                await create(fullAnalysisName, toolLabel, contents)
+                //TODO: Temporary, once Analyses.js uses store, refreshAnalyses will be deprecated in favor of refresh() within the create function
                 await refreshAnalyses()
+                // @ts-expect-error
                 await Ajax().Metrics.captureEvent(Events.analysisCreate, { source: toolLabel, application: toolLabel, filename: fullAnalysisName })
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts:366]: I have a useMetrics hook that will make this less awkward. (part of my pending PR)

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 3]
```diff
@@ -71,10 +75,23 @@ export const useAnalysisFiles = (): AnalysisFileStore => {
...
     setAnalyses(existingAnalyses)
   }) as () => Promise<void>
+
+  const create: (fullAnalysisName: any, toolLabel: ToolLabel, contents: any) => Promise<void> = _.flow(
+    // @ts-expect-error
+    withErrorReporting('Error creating analysis files'),
+    Utils.withBusyState(setPendingCreate)
+  )(async (fullAnalysisName: any, toolLabel: ToolLabel, contents: any): Promise<void> => {
+    const workspaceInfo = workspace.workspace
+    isGoogleWorkspaceInfo(workspaceInfo) ?
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: this if/else should slot well into the provider pattern being used by a few folks to decouple the azure vs gcp differences.  I have an AnalysisProvider that might fit the bill in my pending PR

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 4]
```diff
@@ -52,14 +52,18 @@ export interface AnalysisFile {
...
+  create: (fullAnalysisName: String, toolLabel: ToolLabel, contents: String) => Promise<void>
+  pendingCreate: Boolean
 }
 
 export const useAnalysisFiles = (): AnalysisFileStore => {
   const signal = useCancellation()
   const [loading, setLoading] = useState(false)
   const workspace: WorkspaceWrapper = useStore(workspaceStore)
   const [analyses, setAnalyses] = useState<AnalysisFile[]>([])
+  const [pendingCreate, setPendingCreate] = useState(false)
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: if pendingState is typed as LoadedState<true, unknown>, it can express error cases instead of just true/false state.

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 5]
```diff
@@ -71,10 +75,23 @@ export const useAnalysisFiles = (): AnalysisFileStore => {
...
+      await Ajax().Buckets.analysis(workspaceInfo.googleProject, workspaceInfo.bucketName, fullAnalysisName, toolLabel).create(contents) :
+      await Ajax().AzureStorage.blob(workspaceInfo.workspaceId, fullAnalysisName).create(contents)
+    refresh()
+  }) as () => Promise<void>
+
   useEffect(() => {
     refresh()
   }, [workspace]) // eslint-disable-line react-hooks/exhaustive-deps
-  return { refresh, loadedState: { status: loading ? 'Loading' : 'Ready', state: analyses } }
+  return { refresh, create, loadedState: { status: loading ? 'Loading' : 'Ready', state: analyses }, pendingCreate }
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/file-utils.ts:100]: loadedState is only expressing 'Loading' and 'Ready' cases here.  Consider leveraging the new useLoadedData helper hook to express 'Error' cases as well.

File: types/images/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+declare module '*.png' {
+  const value: any
+  export = value;
+}
+
+declare module '*.svg' {
+  const value: any
```
Comment (msilva-broad) [inline types/images/index.d.ts]: I'm pretty sure there is a nice type for this that maps to the svg structure expected inside

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 1 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: I'll add another comment then! It will be useful to either me or Justin.

File: src/pages/workspaces/workspace/analysis/file-utils.ts (hunk 4 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: I played around with it a little bit but couldn't figure out what syntax to use. Could we schedule a meeting?

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts [hunk 7]
```diff
@@ -109,5 +114,61 @@ describe('file-utils', () => {
...
+      await waitForNextUpdate()
+
+      // Act
+      act(async () => {
+        await hookReturnRef.current.create('AnalysisFile', toolLabels.Jupyter, 'myContents')
+      })
+
+      // Assert
+      expect(create).toHaveBeenCalledWith('myContents')
+    })
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/file-utils.test.ts:204]: we might want to test an error in create as well?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.test.ts [hunk 8]
```diff
@@ -122,6 +124,8 @@ describe('AnalysisDuplicator', () => {
     asMockedFn(useAnalysisFiles).mockImplementation(() => ({
       loadedState: { state: fileList, status: 'Ready' },
       refresh: () => Promise.resolve(),
+      create: () => Promise.resolve(), //TODO: Investigate test
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.test.ts]: this TODO seems done

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts (hunk 7 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/file-utils.test.ts:204]: Just pushed it! It required some adjustment to file-utils.

File: types/images/index.d.ts (hunk 6 above)
Comment (n8landolt) [reply types/images/index.d.ts]: Did you find anything with this?
Is this something @nawatts knows about?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    await userEvent.click(button)
+
+    // Assert
+    screen.getByText(expectedTitle)
+  })
+
+  it.each([
+    { fileType: 'Python 3' },
+    { fileType: 'R' },
+  ])('GCP - Creates a new $fileType for Jupyter when no apps or runtimes are present and navigates to environment creation.', async ({ fileType }) => {
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: nit: Maybe rephrase this slightly from `navigates to environment creation.` to `opens environment creaton modal`. I was looking for asserts around `nav.goToxxx` until I realized what this meant

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    expect(screen.queryByText('Jupyter Cloud Environment')).toBeNull()
+    expect(createFunc).toHaveBeenCalled()
+  })
+
+  it.each([
+    { fileType: 'R Markdown (.Rmd)' },
+    { fileType: 'R Script (.R)' }
+  ])('GCP - Creates a new $fileType for RStudio when no apps or runtimes are present and navigates to environment creation.', async ({ fileType }) => {
+    // Arrange
+    render(h(AnalysisModal, defaultGcpModalProps))
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:216]: nit: i think render is technically part of act, arrange is mainly for variables

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+      const createButton = await screen.findByText('Create Analysis')
+      await userEvent.click(createButton)
+    })
+
+    // Assert
+    screen.getByText('Azure Cloud Environment')
+    expect(createFunc).toHaveBeenCalled()
+  })
+
+  // ----- Work in progress ----
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: Are these still WIP?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 1 above)
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Hmm... is this entirely true though? I think we will still need to bring in the entire `AnalysisFileStore`, even iif iit is from the parent, since a spinner in this file depends on the `loadedState` and the `pendingCreate`. With this in mind, I  think we can modify the above TODO to reflect this, and remove the below TODO

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 12]
```diff
@@ -128,8 +153,9 @@ export const AnalysisModal = withDisplayName('AnalysisModal')(
     })
 
     const renderAzureModal = () => h(AzureComputeModalBase, {
-      isOpen: currentTool === toolLabels.JupyterLab,
-      workspace,
+      workspace: {
+        workspace: workspace.workspace
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: is this needed? it seems like the `workspace` trivially de-structures to this if `workspace: workspace.workspace` is a valid property access

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 13]
```diff
@@ -209,30 +239,48 @@ export const AnalysisModal = withDisplayName('AnalysisModal')(
...
-        accept: `.${tools.Jupyter.ext.join(', .')}, .${tools.RStudio.ext.join(', .')}`,
+        accept: `.${runtimeTools.Jupyter.ext.join(', .')}, .${runtimeTools.RStudio.ext.join(', .')}`,
         style: { flexGrow: 1, backgroundColor: colors.light(), height: '100%' },
         activeStyle: { backgroundColor: colors.accent(0.2), cursor: 'copy' },
         onDropRejected: () => reportError('Not a valid analysis file',
-          `The selected file is not one of the supported types: .${tools.Jupyter.ext.join(', .')}, .${tools.RStudio.ext.join(', .')}. Ensure your file has the proper extension.`),
+          `The selected file is not one of the supported types: .${runtimeTools.Jupyter.ext.join(', .')}, .${runtimeTools.RStudio.ext.join(', .')}. Ensure your file has the proper extension.`),
         onDropAccepted: files => {
-          const tool = !!googleProject ? tools[getToolFromFileExtension(files.pop().path)] : runtimeTools.JupyterLab
+          let tool
```
Comment (jdcanas) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: We really try to stay away from mutation in the codebase...

Could this instead read 
```
const tool = isGoogleWorkspaceInfo(workspace.workspace) ? _.get(getToolFromFileExtension(files.pop().path), tools) : runtimeTools.JupyterLab
```

If so, it might be worth renaming `getToolFromFileExtension` to `getToolLabelFromFileExtension`

File: types/images/index.d.ts (hunk 6 above)
Comment (nawatts) [reply types/images/index.d.ts]: I think the default export should be `string` instead of `any`. The default export is the URL for the image generated by webpack. SVG files also have a `ReactComponent` named export.

**Review by jdcanas — COMMENTED** [review]
First pass. I want to take another look at the tests themselves soon. Mainly just a bit of TODO cleanup needed

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 11 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: Nope!

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 10 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:216]: My thinking was in other tests we adjust some of the default values, so it's essentially setting up the initial render. In my mind, `defaultGcpModalProps` is essentially the "arrange"

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 9 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: Good feedback!

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 10 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:216]: @msilva-broad likely has opinions on this.

It depends on the test. If the test is checking that something is rendered, render could be the "act". On the other hand, if the test is checking some user interaction after the initial render, render may be part of the "arrange".

File: types/images/index.d.ts (hunk 6 above)
Comment (n8landolt) [reply types/images/index.d.ts]: String worked for me!

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 13 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: I renamed a few of the incorrectly named functions from tool-utils.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 10 above)
Comment (msilva-broad) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:216]: nick is correct.  //Arrange is not just about instantiating and setting up mocks... its to get the code into the "pre-condition" state before the main event.... so Nick's examples are spot on.

**Review by msilva-broad — APPROVED** [review]
looks good so far (relying on Justin for IA domain knowledge).
improvements waiting on my PR can wait if the timing doesn't work out.

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts [hunk 14]
```diff
@@ -1,8 +1,10 @@
 import '@testing-library/jest-dom'
 
 import { renderHook } from '@testing-library/react-hooks'
+import { act } from 'react-dom/test-utils'
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.test.ts]: `act` can be imported from `@testing-library/react-hooks`.

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts [hunk 15]
```diff
@@ -109,5 +122,113 @@ describe('file-utils', () => {
...
+        analysis: analysisMock as GoogleStorageContract['analysis']
+      })
+      asMockedFn(GoogleStorage).mockImplementation(() => googleStorageMock as GoogleStorageContract)
+
+      // Act
+      await act(async () => {
+        workspaceStore.set(defaultGoogleWorkspace)
+        const { result: hookReturnRef, waitForNextUpdate } = renderHook(() => useAnalysisFiles())
+        await waitForNextUpdate()
+        await hookReturnRef.current.create('AnalysisFile', toolLabels.Jupyter, 'myContents')
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.test.ts:145]: It looks like this is the only line that needs to be wrapped in `act`. Same comment for other tests in this file.

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 16]
```diff
@@ -52,14 +53,18 @@ export interface AnalysisFile {
 
 export interface AnalysisFileStore {
   refresh: () => Promise<void>
-  loadedState: LoadingState<AnalysisFile[]> | ReadyState<AnalysisFile[]>
+  loadedState: LoadedState<AnalysisFile[], unknown>
+  create: (fullAnalysisName: String, toolLabel: ToolLabel, contents: String) => Promise<void>
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: `string` should be lowercase. Capitalized `String` is a different type.
https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#the-primitives-string-number-and-boolean

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 17]
```diff
@@ -52,14 +53,18 @@ export interface AnalysisFile {
...
+  pendingCreate: LoadedState<true, unknown>
 }
 
 export const useAnalysisFiles = (): AnalysisFileStore => {
   const signal = useCancellation()
   const [loading, setLoading] = useState(false)
   const workspace: WorkspaceWrapper = useStore(workspaceStore)
   const [analyses, setAnalyses] = useState<AnalysisFile[]>([])
+  const [pendingCreate, setPendingCreate] = useLoadedData<true>()
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts:66]: Maybe something like `loadPendingCreate` instead of `setPendingCreate`? When I read, `setSomething(...)`, I think it's setting a state variable. @msilva-broad thoughts on conventions for `useLoadedData`?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts [hunk 18]
```diff
@@ -39,7 +39,9 @@ export interface AnalysisDuplicatorProps {
 
 export const AnalysisDuplicator = ({ destroyOld = false, fromLauncher = false, printName, toolLabel, workspaceInfo, onDismiss, onSuccess }: AnalysisDuplicatorProps) => {
   const [newName, setNewName] = useState<string>('')
-  const { loadedState: { state: analyses } }: AnalysisFileStore = useAnalysisFiles()
+  const { loadedState }: AnalysisFileStore = useAnalysisFiles()
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts]: Probably a matter of style, but the type doesn't really need to be specified here. It's inferred from the return type of `useAnalysisFiles`.

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts [hunk 19]
```diff
@@ -18,20 +20,31 @@ import {
...
 jest.mock('src/libs/ajax/AzureStorage')
+jest.mock('src/libs/error', () => ({
+  ...jest.requireActual('src/libs/error'),
+  reportError: jest.fn(),
+}))
+
+jest.mock('src/libs/notifications', () => ({
+  notify: jest.fn((...args) => {
+    console.debug('######################### notify')/* eslint-disable-line */
+    console.debug({ method: 'notify', args: [...args] })/* eslint-disable-line */
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.test.ts:36]: Not necessary for this PR, but I've seen this same mock in multiple files. Time for a shared mock?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+    console.debug('######################### notify')/* eslint-disable-line */
+    console.debug({ method: 'notify', args: [...args] })/* eslint-disable-line */
+  })
+}))
+
+jest.mock('src/pages/workspaces/workspace/analysis/file-utils', () => {
+  const originalModule = jest.requireActual('src/pages/workspaces/workspace/analysis/file-utils')
+  return {
+    ...originalModule,
+    // getExtension: jest.fn().mockImplementation((ext: Extension) => originalModule.getExtension(ext)), //() => 'ipynb' as Extension,
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: Can this be removed?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 21]
```diff
@@ -0,0 +1,397 @@
...
+  beforeEach(() => {
+    // Arrange
+    asMockedFn(useAnalysisFiles).mockImplementation(() => ({
+      refresh: () => Promise.resolve(),
+      loadedState: { state: [], status: 'Ready' },
+      create: createFunc,
+      pendingCreate: { status: 'Ready', state: true }
+    }))
+
+    //@ts-expect-error
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:92]: What's the error that has to be ignored here?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 22]
```diff
@@ -0,0 +1,397 @@
...
+  })
+
+  it('GCP - Successfully resets view.', async () => {
+    // Act
+    render(h(AnalysisModal, defaultGcpModalProps))
+
+    // Act
+    const button = screen.getByAltText('Create new notebook')
+
+    await userEvent.click(button)
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts]: Instead of calling methods directly from `userEvent`, the docs for `@testing-library/user-event` suggest something like:

```ts
const user = userEvent.setup()
...
await user.click(...)
```

> Note that, while directly invoking APIs such as userEvent.click() (which will trigger setup internally) is [still supported in v14](https://testing-library.com/docs/user-event/setup#direct-apis), this option exists to ease the migration from v13 to v14, and for simple tests. We recommend using the methods on the instances returned by userEvent.setup().

https://testing-library.com/docs/user-event/intro#writing-tests-with-userevent

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 23]
```diff
@@ -0,0 +1,397 @@
...
+    expect(screen.queryByText('RStudio Cloud Environment')).toBeNull()
+    expect(createFunc).toHaveBeenCalled()
+  })
+
+  it('GCP - Renders Galaxy Environment page when no runtime exists and Galaxy is selected.', async () => {
+    // Arrange
+    render(h(AnalysisModal, defaultGcpModalProps))
+
+    // Act
+    await act(async () => {
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:272]: `act` doesn't seem to be needed here.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 24]
```diff
@@ -18,46 +18,73 @@ import { FormLabel } from 'src/libs/forms'
...
 
 const titleId = 'analysis-modal-title'
 const analysisMode = Symbol('artifact')
 const environmentMode = Symbol('environment')
 
+export interface AnalysisModalProps {
+  isOpen: boolean
+  workspace: BaseWorkspace
+  location: any
+  runtimes: any
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Could some of these props be more specifically typed as arrays (`any[]`)?

Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Also might be useful to define a placeholder type in runtime-utils (`type Runtime = any`).

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 1 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: This is one thing I dislike about LoadedState. Many data fetching hooks start loading immediately. In those cases, NoneState isn't used and requires jumping through these hoops to access state.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts [hunk 25]
```diff
@@ -247,60 +286,59 @@ export const AnalysisModal = withDisplayName('AnalysisModal')(
...
+          ]),
+        isRStudio && h(Fragment, [
+          h(FormLabel, { htmlFor: fileTypeSelect, required: true }, ['File Type']),
+          h(Select, {
+            id: fileTypeSelect, isSearchable: true,
+            value: fileExt,
+            onChange: v => {
+              setFileExt(v.value)
+            },
+            options: toolExtensionDisplay.RStudio as ExtensionDisplay[]
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Is this cast necessary? Isn't `toolExtensionDisplay.RStudio` already typed as `ExtensionDisplay[]`?

**Review by nawatts — CHANGES_REQUESTED** [review]
A few minor comments. I think the only things I'd request changes over are a few type corrections (`string` instead of `String`) and getting rid of the notification debug logs in test output.

I feel compelled to once again pontificate about small, focused PRs. It would have been nice to have some of these changes (especially the rename that touched many files) in separate PRs.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 24 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: I would definitely make that placeholder, but it requires a .ts file, currently runtime-utils is .js

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 25 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: It's unhappy about the possibility of being undefined.

What should I do in this situation? Is toolExtensionDisplay improperly typed?

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 24 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Ah. I lost track of which files had been converted. Doesn't necessarily have to be in that file though.

**Comment by n8landolt** [comment]
@nawatts 
`I feel compelled to once again pontificate about small, focused PRs. It would have been nice to have some of these changes (especially the rename that touched many files) in separate PRs.`

I agree - that's my mistake to rename the util functions in this PR. More than happy to revert that change and make a new ticket if you would prefer.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 25 above)
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: Ooh, since it's Partial (`Partial<Record<ToolLabel, ExtensionDisplay[]>>`). I'm not sure.

In that case, I think this could be written more concisely as `toolExtensionDisplay.RStudio!`.

**Comment by nawatts** [comment]
> More than happy to revert that change and make a new ticket if you would prefer.

I won't request that. Just meant it as a reminder.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts (hunk 24 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.ts]: I just threw up a few types in tool-utils.ts with a comment.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 26]
```diff
@@ -0,0 +1,410 @@
...
+    screen.getByText('Select an application')
+    screen.getByAltText('Create new notebook')
+    screen.getByAltText('Create new R file')
+    screen.getByAltText('Create new Galaxy app')
+  })
+
+  it('GCP - Successfully resets view.', async () => {
+    // Arrange
+    render(h(AnalysisModal, defaultGcpModalProps))
+    const user = userEvent.setup()
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:108]: > We recommend invoking [userEvent.setup()](https://testing-library.com/docs/user-event/setup) before the component is rendered.

https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent

**Review by nawatts — COMMENTED** [review]
Seeing one warning from AnalysisModalTest about an update not wrapped in act.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts (hunk 26 above)
Comment (n8landolt) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:108]: Oops! I had a feeling the order was important. Fixing.

File: src/pages/workspaces/workspace/analysis/file-utils.test.ts [hunk 27]
```diff
@@ -109,5 +121,113 @@ describe('file-utils', () => {
...
+      })
+      asMockedFn(GoogleStorage).mockImplementation(() => googleStorageMock as GoogleStorageContract)
+
+      // Act
+      workspaceStore.set(defaultGoogleWorkspace)
+      const { result: hookReturnRef, waitForNextUpdate } = renderHook(() => useAnalysisFiles())
+      await waitForNextUpdate()
+      await act(async () => {
+        await hookReturnRef.current.create('AnalysisFile', toolLabels.Jupyter, 'myContents')
+      })
```
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.test.ts:146]: Not necessary to change, but one line `act`s can be written more concisely as:

```suggestion
      await act(() => hookReturnRef.current.create('AnalysisFile', toolLabels.Jupyter, 'myContents'))
```

**Review by nawatts — APPROVED** [review]

**Review by jdcanas — APPROVED** [review]
